### ISSUE TITLE
Clarification what shaderInt16 and shaderInt64 actually mean

### ISSUE THREAD

**Issue by Novum** [KhronosGroup/Vulkan-Docs#715]
Related issue on glslang side: https://github.com/KhronosGroup/glslang/issues/1373

The spec states:
"shaderInt16 specifies whether 16-bit integers (signed and unsigned) are supported in shader code. If this feature is not enabled, 16-bit integer types must not be used in shader code. This also specifies whether shader modules can declare the Int16 capability."

What does "use" *actually* mean? Does it allow the use of int16 for arithmetic? For loads and stores?

This is especially confusing since there is VK_AMD_gpu_shader_int16 which specifies the use for arithmetic and VK_KHR_16bit_storage which allows load/store.

If there are extensions for both is shaderInt16 completely useless, because neither are allowed by base Vulkan? Am I missing something?

**Comment by krOoze** [comment]
So, I would say "use" means what the 'Table 75. List of optional SPIR-V Capabilities and corresponding feature or extension names' in 'Appendix A: Vulkan Environment for SPIR-V' says. I.e. that you can pass a shader program to Vulkan that has the `Int16`\`Int64`\`Float64` SPIR-V capability declared.

`VK_KHR_16bit_storage` should not theoretically need it. It has its own set of SPIR-V capabilities and special rules.

Literal reading of `VK_AMD_gpu_shader_int16` (resp. `SPV_AMD_gpu_shader_int16`) says it just enables `i16` type use in two GLSL instruction set functions. It does **not** influence whether you can create an `i16` in the first place (i.e. whether `Int16` capability was declared). It makes sense to have that as an SPIR-V extension, because vanilla GLSL does not even have `i16`.
But people (incl. some AMD employees) seem to treat similar extension `VK_AMD_gpu_shader_half_float` as if it allows `Float16` capability (#706 ), so IDK...

**Comment by Novum** [comment]
So shaderInt16 literally means nothing than I can declare an int16 in SPIR-V, but do absolutely nothing with it? Why was this even added?

**Comment by krOoze** [comment]
What do you mean by "absolutly nothing"? You can use it as an argument to any instruction that takes `OpTypeInt` and does not have any other restrictions.

**Comment by krOoze** [comment]
> So shaderInt16 literally means that I can declare an int16 in SPIR-V, but do absolutely nothing with it?
That brings me to the question what the intention of this whole thing was?

If you are so inclined, but you (resp. your compiler) would probably not do nothing with it.
The intention is probably to be able to easily reflect a SPIR-V binary without having to parse it whole for uses of `int16`.

**Comment by Novum** [comment]
But then again, just declaring the value without being able to do *anything* with it can be supported on any GPU. Because there is no actual code generated for it.

**Comment by Novum** [comment]
It's surprising because there is literally no use for it. All functionality that requires int16 is in extensions and those already require OpType int16.

This just causes confusion, e.g. AMD drivers report VK_KHR_16bit_storage, but not shaderInt16. Which makes zero sense. They probably assume shaderInt16 actually provides functionality - which it doesn't.

**Comment by krOoze** [comment]
@Novum  OK, no, there is some division of labor between all those.

`shaderInt16` feature (resp. `Int16` cap) allows to define `int16_t` type (from which follows it can then be used as parameters in SPIR-V ops).

`VK_KHR_16bit_storage` allows creating pointers to `int16_t`. They may be converted to `int32`, but without the above cap not otherwisely used.

So:
>  e.g. AMD drivers report `VK_KHR_16bit_storage`, but not `shaderInt16`. Which makes zero sense.

actually makes perfect sense. The impl can load 16-bit from memory. But it probably does not have native 16-bit ALUs (and would have to emulate `int16_t` arithmetics, which you likely do not want, so it makes you use them only as `int32`).

**Comment by Novum** [comment]
No. OpTypeInt 16 is needed for the load in SPIR-V. So you need to declare OpCapability Int16, therefore the driver needs to advertise shaderInt16.

**Comment by krOoze** [comment]
Why would you think it does not provide any ALU ops?? It does(, with only some rare exceptions that state they need 32-bit int specifically).

You cannot just ignore uses of `int16_t` in the shader. E.g., you could load `int32_t` and unpack it into two `int16_t`s (and so you would not be able to get rid of that by constant folding, or by removing dead flow).

> No. OpTypeInt 16 is needed for just the load without arithmetic in SPIR-V. So you need to declare OpCapability Int16, therefore the driver needs to advertise shaderInt16.

You do not. Read [SPV_KHR_16bit_storage](https://www.khronos.org/registry/spir-v/extensions/KHR/SPV_KHR_16bit_storage.html). It allows declaring `OpTypeInt 16` **conditionally** even without `Int16` cap.

So, driver does not need to either.

> It's especially inconsistent since shaderFloat16 is missing.

Indeed. That's what Issue #706 is all about.

> [...], but you can't actually do anything with it without additional extensions.

You can, as demonstrated above. With `SPV_AMD_gpu_shader_int16` you only get three extra instructions accepting `int16_t`. With `SPV_KHR_16bit_storage` you can load `int16_t` directly (without having to pack\unpack). Rest of it (I would say the meat) is done by the `Int16` cap (resp. `shaderFloat16` feature).

> I'm very sure people assume a)

With the above said, it might mean "people" were right all along?

**Comment by Novum** [comment]
I had that impression, because the SPIR-V spec just says "Int16 UsesOpTypeInt to declare 16-bit integer types". There is no mention of anything else. Also the GLSL compiler insists of adding SPV_AMD_gpu_shader_int16 when doing an 16 bit integer ops, which I interpretes as "there are no int16 ops without the extension".

**Comment by krOoze** [comment]
> Also the GLSL compiler insists of adding `SPV_AMD_gpu_shader_int16` when doing an 16 bit integer ops

That is kinda bad, because almost no GPU has that extension.

> If it is indeed the case that shaderInt16 provides ALU ops, then there is probably a bug in glslang.

The "correct"\desirable behavior is only gated to Vulkan 1.1 (resp. SPIR-V 1.3) in the glslang code for some reason (or lack thereof).

**Comment by alecazam** [comment]
I really don't understand Vulkan float16_t storage vs. shader support.  Splitting this functionality is useless as Novum details.  It's even worse in GLSL where on Android and on consoles that are missing shaderFloat16, it's near impossible to efficiently write shaders that use half.   

Then throw on top that Nvidia doesn't support storage.inputOutput16, so you can't get the benefit on TBDR systems of saving parameter buffer system memory by outputting half to the fragment shader.  And if the storage16 extension isn't present, then it's even wors.e

https://github.com/KhronosGroup/SPIRV-Tools/issues/4546
